src/combinator.rs

```diff
@@ -0,0 +1,305 @@
+            f(&x).0.perform(Context { i, o, e })?;
+            Ok (x)
+        })
+    }
+
+    /// Like [`tap`](IO::tap), but "impure": its side effects happen
+    /// outside of the [`IO`] monad. This has some niche uses, for
+    /// example [`trace`](IO::trace) is implemented in terms of
+    /// `tap_impure`.
+    pub fn tap_impure <F> (self, f: F) -> IO<'a, T>
+    where
+        F: for<'b> FnOnce (&'b T) + 'a,
+    {
+        self.tap(|x| pure(f(x)))
+    }
+
+    /// Write `msg` to stderr *directly*, leaving the value untouched.
+    ///
+    /// The value written to stderr will *not* be captured in any way.
+    ///
+    /// ```
+    /// use iom::{ IO, empty };
+    ///
+    /// # fn main () -> std::io::Result<()> {
+    /// let chain = empty.trace("I'm").trace("spamming");
+    /// let result = IO::capture(chain)?;
+    ///
+    /// assert!(result.is_empty());
+    /// # Ok (()) }
+    /// ```
+    pub fn trace (self, msg: impl Display + 'a) -> IO<'a, T> {
+        self.tap_impure(move |_| {
+            // Bypass the IO monad altogether when writing to stderr.
+            // This makes it so the message cannot be captured.
+            eprintln!("{msg}")
+        })
+    }
+
+    /// Apply a function to the wrapped value.
+    ///
+    /// `map` on [`IO`] behaves similar to [`Option::map`], [`Result::map`]
+    /// or [`Iterator::map`]. After this IO action is performed, its result
+    /// is passed to the closure `f` and its output is then returned.
+    pub fn map <U, F> (self, f: F) -> IO<'a, U>
+    where
+        F: FnOnce (T) -> U + 'a,
+        U: 'a,
+    {
+        self.bind(|x| pure(f(x)))
+    }
+
+    /// Replace the value currently in the monad with a new `value`, discarding
+    /// the old one.
+    ///
+    /// All actions are still performed. This is equivalent to `self.map(|_| value)`.
+    pub fn replace <U> (self, value: U) -> IO<'a, U>
+    where
+        U: 'a,
+    {
+        self.map(|_| value)
+    }
+
+    /// Perform `self` and discard the output, then perform `other` and return
+    /// its output.
+    pub fn then <'b: 'a, U> (self, other: IO<'a, U>) -> IO<'a, U>
+    where
+        U: 'b,
+    {
+        self.bind(|_| other)
+    }
+
+    /// Perform `other` after `self`, but ignore its output and return `self`'s
+    /// output instead.
+    pub fn discard <U> (self, other: IO<'a, U>) -> IO<'a, T>
+    where
+        U: 'a,
+    {
+        self.bind(|v| other.replace(v))
+    }
+
+    /// Read `self`'s output and pass it to the closure `f`.
+    ///
+    /// ```
+    /// use iom::{ IO, println, pure, empty };
+    ///
+    /// # fn main () -> std::io::Result<()> {
+    /// // `b"meow"` is an array of chars, so it
+    /// // does not implement `Read`.
+    /// let buf = b"meow".as_ref();
+    /// let a = pure(buf).with_content(|s| {
+    ///     assert_eq!(s, "meow");
+    ///     empty
+    /// });
+    ///
+    /// IO::run(a)?;
+    /// # Ok (()) }
```

Comment:
This example should probably use assert_eq instead of a comment

Reply:
I'm not sure what you mean, can you explain further?

Reply:
If you do
```suggestion
    /// assert!(matches!(IO::run(a)?, Ok(()))
```

it's most in line with how rust docs tend to do things and allows doctests to check if IO::run really returns Ok(())

Reply:
It needs to return `Ok (())` at the end because it is wrapped in `fn main () -> std::io::Result<()>` though. It's not a comment, rather the line with the `Ok (())` is hidden.